src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -764,5 +1551,177 @@ private static bool ShouldRoundUp(bool lsbBit, bool roundBit, bool hasTailBits)
+                // Suppose we start with 2.2250738585072013e-308, we end up
+                // with 0x3fffffffffffff x 2^-1023-53 which is technically subnormal
+                // whereas 0x40000000000000 x 2^-1023-53  is normal. Now, we need to round
+                // up 0x3fffffffffffff x 2^-1023-53  and once we do, we are no longer
+                // subnormal, but we can only know this after rounding.
+                // So we only declare a subnormal if we are smaller than the threshold.
+                answer.power2 = (answer.mantissa < ((ulong)(1) << info.DenormalMantissaBits)) ? 0 : 1;
+                return answer;
+            }
+
+            // usually, we round *up*, but if we fall right in between and and we have an
+            // even basis, we need to round down
+            // We are only concerned with the cases where 5**q fits in single 64-bit word.
+            if ((product.low <= 1) && (q >= info.MinExponentRoundToEven) && (q <= info.MaxExponentRoundToEven) &&
+                ((answer.mantissa & 3) == 1))
+            { // we may fall between two floats!
+              // To be in-between two floats we need that in doing
+              //   answer.mantissa = product.high >> (upperbit + 64 - mantissa_explicit_bits() - 3);
+              // ... we dropped out only zeroes. But if this happened, then we can go back!!!
+                if ((answer.mantissa << (upperbit + 64 - info.DenormalMantissaBits - 3)) == product.high)
+                {
+                    answer.mantissa &= ~(ulong)(1);          // flip it so that we do not round up
+                }
+            }
+
+            answer.mantissa += (answer.mantissa & 1); // round up
+            answer.mantissa >>= 1;
+            if (answer.mantissa >= ((ulong)(2) << info.DenormalMantissaBits))
+            {
+                answer.mantissa = ((ulong)(1) << info.DenormalMantissaBits);
+                answer.power2++; // undo previous addition
+            }
+
+            answer.mantissa &= ~((ulong)(1) << info.DenormalMantissaBits);
+            if (answer.power2 >= info.InfinitePower)
+            { // infinity
+                answer.power2 = info.InfinitePower;
+                answer.mantissa = 0;
+            }
+            return answer;
+        }
+        private static value128 ComputeProductApproximation(int bitPrecision, long q, ulong w)
+        {
+            int index = 2 * (int)(q - SmallestPowerOfFive);
+            // For small values of q, e.g., q in [0,27], the answer is always exact because
+            // FullMultiplication gives the exact answer.
+            value128 firstproduct = FullMultiplication(w, s_Pow5128Table[index]);
+            ulong precision_mask = (bitPrecision < 64) ? ((ulong)(0xFFFFFFFFFFFFFFFF) >> bitPrecision) : (ulong)(0xFFFFFFFFFFFFFFFF);
+            if ((firstproduct.high & precision_mask) == precision_mask)
+            { // could further guard with  (lower + w < lower)
```

Comment:
`{ //` here and in other places could you please start a new line for comments?
```
{ // comment
vs
{
    // comment
```

Reply:
done.